src/gui/qgssnaptogridcanvasitem.cpp

```diff
@@ -0,0 +1,166 @@
+    ----------------------
+    begin                : August 2018
+    copyright            : (C) Matthias Kuhn
+    email                : matthias@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgssnaptogridcanvasitem.h"
+#include "qgsmapcanvas.h"
+
+QgsSnapToGridCanvasItem::QgsSnapToGridCanvasItem( QgsMapCanvas *mapCanvas )
+  : QgsMapCanvasItem( mapCanvas )
+  , mGridPen( QPen( QColor( 127, 127, 127, 150 ), 1 ) )
+  , mCurrentPointPen( QPen( QColor( 200, 200, 200, 150 ), 3 ) )
+{
+  updateMapCanvasCrs();
+  connect( mMapCanvas, &QgsMapCanvas::extentsChanged, this, &QgsSnapToGridCanvasItem::updateZoomFactor );
+  connect( mMapCanvas, &QgsMapCanvas::destinationCrsChanged, this, &QgsSnapToGridCanvasItem::updateMapCanvasCrs );
+}
+
+void QgsSnapToGridCanvasItem::paint( QPainter *painter )
+{
+  painter->save();
+  QgsRectangle mapRect = mMapCanvas->extent();
+  if ( rect() != mapRect )
+    setRect( mapRect );
+
+  painter->setRenderHints( QPainter::Antialiasing );
+  painter->setCompositionMode( QPainter::CompositionMode_Difference );
+
+  if ( mEnabled && mAvailableByZoomFactor )
+  {
+    try
+    {
+      const QgsRectangle layerExtent = mTransform.transformBoundingBox( mapRect, QgsCoordinateTransform::ReverseTransform );
+      const QgsPointXY layerPt = mTransform.transform( mPoint, QgsCoordinateTransform::ReverseTransform );
+
+      const double gridXMin = std::ceil( layerExtent.xMinimum() / mPrecision ) * mPrecision;
+      const double gridXMax = std::ceil( layerExtent.xMaximum() / mPrecision ) * mPrecision;
+      const double gridYMin = std::ceil( layerExtent.yMinimum() / mPrecision ) * mPrecision;
+      const double gridYMax = std::ceil( layerExtent.yMaximum() / mPrecision ) * mPrecision;
+
+      for ( int x = gridXMin ; x < gridXMax; x += mPrecision )
+      {
+        for ( int y = gridYMin ; y < gridYMax; y += mPrecision )
+        {
+          const QgsPointXY pt = mTransform.transform( x, y );
+          const QPointF canvasPt = toCanvasCoordinates( pt );
+
+          if ( qgsDoubleNear( layerPt.x(), x, mPrecision / 3 ) && qgsDoubleNear( layerPt.y(), y, mPrecision / 3 ) )
+          {
+            painter->setPen( mCurrentPointPen );
+          }
+          else
+          {
+            painter->setPen( mGridPen );
+          }
+          painter->drawLine( canvasPt.x() - 3, canvasPt.y(), canvasPt.x() + 3, canvasPt.y() );
```

Comment:
The `3` constants here aren't hi-dpi friendly

Reply:
Use something font size dependant. E.g. `fontMetrics()->width( 'X')`

Reply:
Does the pen width (see initializer list, where you proposed to move to header) need to be adapted too? I guess so ...

Reply:
Yes. Unfortunately 0 width cosmetic pens don't scale for hidpi, which leaves the only option manually adjusting. We have this problem with most map tools incidentally - the rubber bands can be almost invisible on hidpi. 

I think the best way to handle this would be some setting QgsGui::isHiDpi - and if it's not, use a nice-looking 0 width cosmetic pen. If it is, use ... I dunno, the width of a full stop?

Reply:
I've implemented something that looks good on hidpi, will need a test by a non-hi-dpi device to see the result.